botan/tests/tests.rs

```diff
@@ -913,3 +913,62 @@ fn test_totp() -> Result<(), botan::Error> {
     assert!(!totp.check(90693936, 59 + 31, 1)?);
     Ok(())
 }
+
+#[cfg(feature = "botan3")]
+#[test]
+fn test_srp6() -> Result<(), botan::Error> {
+    const IDENTITY: &str = "alice";
+    const PASSWORD: &str = "password123";
+    let mut rng = botan::RandomNumberGenerator::new_system()?;
+
+    // Test successful authentication
+    let mut server = botan::ServerSession::new()?;
+    let salt = rng.read(24)?;
+    let verifier =
+        botan::generate_srp6_verifier(IDENTITY, PASSWORD, &salt, "modp/srp/1024", "SHA-512")?;
+    let b_pub = server.step1(&verifier, "modp/srp/1024", "SHA-512", &rng)?;
+    let (a_pub, client_key) = botan::srp6_client_agree(
+        IDENTITY,
+        PASSWORD,
+        "modp/srp/1024",
+        "SHA-512",
+        &salt,
+        &b_pub,
+        &rng,
+    )?;
+    let server_key = server.step2(&a_pub)?;
+    assert_eq!(client_key, server_key);
+
+    // Test wrong server's B value
+    let salt = rng.read(24)?;
+    let b = b"BD0C6151 2C692C0C B6D041FA 01BB152D 4916A1E7 7AF46AE1 05393011 \
+    BAF38964 DC46A067 0DD125B9 5A981652 236F99D9 B681CBF8 7837EC99 \
+    6C6DA044 53728610 D0C6DDB5 8B318885 D7D82C7F 8DEB75CE 7BD4FBAA \
+    37089E6F 9C6059F3 88838E7A 00030B33 1EB76840 910440B1 B27AAEAE \
+    EB4012B7 D7665238 A8E3FB00 4B117B58";
+    let result = botan::srp6_client_agree(
+        IDENTITY,
+        PASSWORD,
+        "modp/srp/1024",
+        "SHA-512",
+        &salt,
+        b,
+        &rng,
+    );
+    assert!(result.is_err());
+
+    // Test wrong client's A value
+    let salt = rng.read(24)?;
+    let verifier =
+        botan::generate_srp6_verifier(IDENTITY, PASSWORD, &salt, "modp/srp/1024", "SHA-512")?;
+    let _ = server.step1(&verifier, "modp/srp/1024", "SHA-512", &rng)?;
+    let a_pub = b"61D5E490 F6F1B795 47B0704C 436F523D D0E560F0 C64115BB 72557EC4 \
```

Comment:
The patch here looks good but made me realize something I completely missed in the review of the FFI patch which was that we were taking the input for B as, effectively, a string in decimal or hex format, rather than as a binary value. Can you create a followup patch for the FFI impl? The patch should basically just change `auto b_bn = Botan::BigInt::decode(b, b_len);` to `Botan::BigInt b_bn(b, b_len)` plus changing the test inputs.

Reply:
Sure, will do.

Reply:
Wait, I'm not sure I understand. My patch is taking binary values already:
```   int botan_srp6_client_agree(const char* identity, const char* password,
                               const char* group_id, const char* hash_id,
                               const uint8_t salt[], size_t salt_len,
                               const uint8_t b[], size_t b_len, botan_rng_t rng_obj,  // <--- b is uint8_t[]
                               uint8_t A[], size_t* A_len, uint8_t K[],
                               size_t* K_len)
```
According to the `bitint.h`, this should be the method, right:
```     /**
     * Create a BigInt from an integer in a byte array
     * @param buf the binary value to load
     * @param length size of buf
     * @result BigInt representing the integer in the byte array
     */
     static BigInt decode(const uint8_t buf[], size_t length)
        {
        return BigInt(buf, length);
        }
```

It's probably confusing that in this change I treat the string as a byte slice, but it's not in fact decoded. Can change that.

Reply:
@randombit, please advise :)

Reply:
I think what is confusing me here is this hex-encoded constant in the Python tests https://github.com/randombit/botan/blob/master/src/scripts/test_python.py#L703-L708 which makes it look like the FFI function is expecting a hex encoding.